### PR TITLE
Revert "Refactor operators & MKLDNN (#8302)"

### PR DESCRIPTION
**Pull request by cjolivier01**
This reverts commit 2cc2aa2.
This commit in question disables a large number of unit tests.
## Description ##
(Brief description on what this PR is about)

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by cjolivier01** [comment]
CI C++ unit tests are run by: https://github.com/apache/incubator-mxnet/pull/9609

**Comment by marcoabreu** [comment]
I think we should create a mapping which shows which tests are covered by Python and thus don't need to run in C++ and which are not covered. If everything is covered, I'm fine with removing them. Otherwise, I'd like to see them enabled to prevent reducing test coverage.

Cpp tests will be enabled by tomorrow.

**Comment by cjolivier01** [comment]
Unit tests test correctness of output as well as consistency of *all inputs and outputs* between different implementations (mkl, cpu, gpu, cudnn, etc.)

Unit tests also test varying channel axes.  

In addition, batch norm tests in test_operator.py are disabled due to CI issues, so this leaves batch norm largely untested.

C++ unit tests also measure performance differences between the various versions (native, mkl, gpu, cudnn, etc).

**Comment by marcoabreu** [comment]
See https://github.com/apache/incubator-mxnet/issues/8044 
https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_operator.py#L994

**Comment by TaoLv** [comment]
Disappointed to see the mkl-dnn pr is reverted. I totally agree that we should enable more test cases to take care of both functionality and performance. Maybe we should open a new issue/pr to add those CPP tests back after the CPP CI is enabled. Those test cases need be carefully redesigned to fit new APIs and new features, rather than remove new APIs and features for keeping legacy test cases. Anyway, we will keep working on mkl-dnn integration and providing more tests and validation. Thanks, team.

**Comment by marcoabreu** [comment]
Hello Tao,
sorry for the revert. Unfortunately, the PR has been merged before it passed all reviews and since we have absolutely no test coverage of batchnorm, I don't feel comfortable having a change of this size untested in our master branch. Problem here being the fact that many customers and developers are building off the master branch and thus having such an unclear situation could harm our users. We will ensure to resolve the addressed problems as soon as possible. 

The CPP tests on Ci will be enabled tomorrow. 

If new APIs and features are being introduced with that PR, there should have been tests covering these cases in the first place. Our PR-template clearly states that all new features require test coverage. We'd like to refrain from adding unit tests in a second PR as this leaves the master branch in an untested state in the meantime. This could have a possible impact on other PRs and thus create a bad developer experience.

**Comment by TaoLv** [comment]
@marcoabreu thanks for your explain. Let's move forward to get mkl-dnn back. I am trying to build #8302 with batchnorm_test enabled and working on those compiling issues in it. Do you think those test cases will run into the new mkl-dnn operator if they are enabled, or we need adjust something else for them.

**Comment by cjolivier01** [comment]
the classes in batchnorm_test (validators, etc) probably needs to be
adjusted to use the core op runner as the operator runner rather than
legacy op runner.

**Comment by marcoabreu** [comment]
I'd propose that we re-enable as many tests as possible and write new tests for all newly introduced APIs. Afterwards, I'd be happy to trigger multiple builds in order to catch flaky tests.

**Comment by TaoLv** [comment]
@cjolivier01 Looks a little tricky for me :(  Shall we enable more python test cases for bn to cover these cpp tests or fix the skipped ut at first? If so, can you list the test points here?
Why bn is so different? It seems there is no cpp test case for other nn operators.

**Comment by cjolivier01** [comment]
intent is to keep these tests in c++

**Comment by cjolivier01** [comment]
Here is starter code to convert the unit tests.  As in the notes, I don;t know youe API so some of it is just using arbitrary input/output vectors, but the core op executor is creating them for the op, I expect.

https://github.com/cjolivier01/mxnet/pull/2

**Comment by mli** [comment]
@marcoabreu I'm curious why this PR has been merged before discussion and reaching an agreement.

If I understand correctly, PR #8302 

1. a very large PR has been intensively reviewed 
2. it commented out the batchnorm cpp test due to 
   1. it is not been enabled in CI before this PR is merged.
   2. it uses an old API and deprecated batchnorm-v1
   3. to re-enable these tests need to update the codes, 1.8K C++ codes to just test batchnorm seems an overkill to me
3. @zheng-da pinged @cjolivier01 twice before merging #8302 hope to discuss it, but there is no response

I can imagine @cjolivier01 's frustration for his codes has been commented, but at the same time, reverting a PR that @zheng-da has been working on it for a half year also seriously hurts his feeling. 

I would like to discuss

1. How long we should wait to allow contributors to a particular a thread, especially for large PRs
2. Is reverting a PR a better pratcise than patching based on the merged PR. Currently #9661 reverts #8302, and then #9609 is merged a hour ago to enable cpp tests on CI. Another option is just trying to update the batchnorm test and then enable cpp tests based on #8302

**Comment by cjolivier01** [comment]
@mli 
This is not true that I was pinged with no response. We talked the day before about how the changes would be implemented.  In addition. comments werre made in the PR about making sure new code is tested.

**Comment by mli** [comment]
@cjolivier01 

Thanks for clarifying it. I just started to look into this thread an hour ago, I definitely ignored something. Can you please point it out your comments, and let's figure out how to move forward.

**Comment by cjolivier01** [comment]
My Comment on the PR Dec 8: Please make sure all googletest tests in tests/cpp still build and pass

Another on Dec 8: Are the batch norm unit tests in tests/cpp/operator/batchnorm_test.cc still working?

Neither had any answering comment.

**Comment by mli** [comment]
@cjolivier01 

If I understand correctly: 

1. 2 days ago @cjolivier01 commented that LGTM (I cannot find the comments on Dec 8)
1. 2 days ago @marcoabreu noticed some tests are disabled
1. @zheng-da told me that he pinged @cjolivier01 on chime after that, he noticed @cjolivier01 read the message but didn't response
1. @piiswrong merged the PR 22 hours ago
1. @cjolivier01 replied after the PR has been merged both on github and chime

What I can see is that, before merging #8302, @cjolivier01 may don't have enough time to have a review. Same thing happens for this reverting PR. So I proposed to discuss that, how long we should wait so that everyone is able to have a look the issues?

**Comment by cjolivier01** [comment]
As you can see from the PR, I reviewed this PR before and made comments.

Since we had spoken before about how to fix some of the unit tests before, I assumed he had fixed the unit tests) in accordance with my comments.  I was not aware until Marco pointing it out that he had actually disabled many of the tests, which was before the PR was merged.

I was pinged in the morning, and home with a sick child. I did respond later that day -- I was not informed that the PR was going to be merged as-is.
Regardless of the timeline or anything else, batchnorm -- one of the most commonly used operators in deep learning, for which we have 5+ distinct implementations -- is now not tested at all, which is why the PR was reverted.  It is nothing personal to Da and his fine work in this PR.

**Comment by zheng-da** [comment]
@cjolivier01 I have fixed all other C++ unit tests. This batchnorm unit test is really too large. It seems to me that it requires a lot of efforts to fix it. I tried multiple times and put it away.
@cjolivier01 I guess there was some misunderstanding. I saw you see my message but didn't respond. I thought you didn't want to have the discussion.

**Comment by zheng-da** [comment]
BTW, it seems to me batchnorm is fairly well tested in test_operator_gpu.py.
https://github.com/zheng-da/incubator-mxnet/blob/master/tests/python/gpu/test_operator_gpu.py#L268

**Comment by szha** [comment]
@zheng-da @cjolivier01 Let's take this as an opportunity to improve the batchnorm tests. Working together, you two can probably figure out a better way for testing. Since the existing 1800 lines of batch norm C++ tests do sound a bit hard to maintain, it would be good to figure out whether some of them can be moved elsewhere. Testing correctness of operators ideally shouldn't take that many lines. Also, performance tests should happen elsewhere outside of unit tests.

**Comment by yzhliu** [comment]
Agree that we'd better to make unit tests work. But I'm really surprised to see so many test cases in `batchnorm_test.cc`. They are not quite self-explained to me. What are they trying to cover in addition to those in `test_operator_gpu.py`?

**Comment by kevinthesun** [comment]
Do we have any tests for batchnorm that cannot be covered by python unit test and have to be implemented in c++? If there isn't any, we should just keep python unit test. If there is, we can consider how to refactor the tests, since 1800 lines of test codes means there is a high possibility that we have duplicated tests in python and cpp, which is not very maintainable.

**Comment by marcoabreu** [comment]
Since Chris already responded, I'll skip answering above questions. If you would like me to elaborate anyways, feel free to give me a heads up @mli .

Regarding Python vs C++. I'd like to point out the importance of C++ for edge devices. Many of them are not able to execute Python code and rely on low-level C++ tests. While I totally agree that a high-level Python test is way easier and more transparent to write, there's also a place in our world for native C++ tests.
